src/sources/connect/wasm_validators.rs

```diff
@@ -0,0 +1,89 @@
+//! Validation helpers for WASM targets
+//!
+//! These utility methods are mostly used for allowing WASM contexts to validate
+//! connect-enabled schemas. Since this must pass the WASM FFI boundary, they can
+//! only operate on WASM primitives (for now) like strings.
+
+use crate::{
+    error::FederationError,
+    schema::ValidFederationSchema,
+    sources::connect::spec::schema::{
+        ConnectDirectiveArguments, SourceDirectiveArguments, CONNECT_DIRECTIVE_NAME_IN_SPEC,
+        SOURCE_DIRECTIVE_NAME_IN_SPEC,
+    },
+};
+
+#[derive(Default)]
+pub struct Location {
+    pub line: usize,
+    pub column: usize,
+}
+#[derive(Default)]
+pub struct LocationRange {
+    pub start: Location,
+    pub end: Location,
+}
+pub struct ValidationError {
+    pub range: LocationRange,
+    pub reason: String,
+    pub error: FederationError,
+}
+
+/// Validate a subgraph schema with connect directives
+// Note: The error is boxed because it is large and clippy complains
+pub fn validate_connect_enabled_schema(subgraph_schema: &str) -> Result<(), Box<ValidationError>> {
+    // Helper for mapping into the correct error type
+    fn error_with_phony_loc(reason: &str, e: impl Into<FederationError>) -> Box<ValidationError> {
+        Box::new(ValidationError {
+            range: Default::default(),
+            reason: reason.to_string(),
+            error: e.into(),
+        })
+    }
+
+    // First we need to create an actul schema from the provided one
+    let schema = apollo_compiler::Schema::parse(subgraph_schema, "")
+        .map_err(|e| error_with_phony_loc("failed to parse schema", e))?;
+
+    // Make sure that the parsed schema is actually valid
+    let schema = schema
+        .validate()
+        .map_err(|e| error_with_phony_loc("failed to validate schema", e))?;
+    let schema = ValidFederationSchema::new(schema)
+        .map_err(|e| error_with_phony_loc("failed to parse as federation schema", e))?;
+
+    // Process the @source directives
+    let sources = schema
+        .referencers()
+        .get_directive(&SOURCE_DIRECTIVE_NAME_IN_SPEC)
```

Comment:
this will error if it can't find any sources — before we do anything, we should first check to see if the `@link` is present using `schema.metadata()` (and return early if it's not), and we should ignore the Err case for sources since they're optional.